### DISCUSSION TITLE
linting qmd files

### DISCUSSION THREAD

**Discussion by skwde** [quarto-dev/quarto-cli#9349] [Q&A]
### Description

It seems that there is no linting for `.qmd` files (in VSCode).

Neither for the specific quarto includes nor for pandoc markdown.
It would be nice to have some linting.

Here are common examples:

Error:

```qmd
Both errors in the following

{{< _basics.md >>}}
just goes unnoticed. (apart from a syntax highlighting change)
```

Style / wrong rendering:

```qmd
some list
- is rendered incorrect
-     and is also inconsistent in spacing
```

Note, there is https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint which is based on https://github.com/DavidAnson/markdownlint to lint markdown / commonmark files. This could at least cover partly the markdown aspect.

Is something like this planned? Or even better already existing and I just didn't stumbled across while searching.

**Comment by mcanouil** [comment]
VSCode Quarto extension repository is <https://github.com/quarto-dev/quarto>.

Search if there is no existing issue about this and if not you can open a feature request.

> edit:
> - https://github.com/quarto-dev/quarto/issues/208
> - https://github.com/DavidAnson/markdownlint/issues/1145 (closed as duplicate of https://github.com/DavidAnson/vscode-markdownlint/issues/228)

Note that if you change the language of the file from "quarto" to "markdown", you get the linters.

**Reply by skwde** [reply]
Apparently, to my understanding, this does not cover all of my questions.

What about quarto specific syntax like includes, i.e. `{{< _basics.md >>}}`?

Ok, so the suggestion for linting the markdown is what I actually already do and there is no specific linter for pandoc markdown included.

**Reply by mcanouil** [reply]
The include shortcode you show is invalid and a Markdown linter such as the one you referenced can't lint that because it's a Quarto thing.
The include shortcode is `{{< include _basics.md >}}`.

A linter cannot tell that something does not exist and since anyone can make a shortcode such as `{{< _basics.md >}}` which could display the date if I write a shortcode filter in Lua with that name and for that purpose.

Quarto will show you an error `Shortcode '_basics.md' not found`.

While if the file you want't to include does not exist:

```sh
ERROR: Include directive failed.
  in file /path/to/file/index.qmd, 
  could not find file /path/to/file/_bad.qmd.
```

---

All this being said, here the thing that is wrong (in my opinion) is that no errors or warnings are raised when rendering (the "shortcode" is simply "removed", not part of the AST):

```qmd
`{{< include _basics.md >>>>>>>>>>>>>>}}`
```

Cc @cscheid

**Reply by skwde** [reply]
Of course you are right I only realized now, that I forgot the `include` in `{{< include _basics.md >>}}` above (sorry about that), but the result and my point stays exactly the same.

I am aware that no markdown linter will handle that and that is also exactly the issue here.

> A linter cannot tell that something does not exist 

Why not? If something is not in path it does not exist?!
E.g. in python linters warn when undefined variables are detected.

It would be nice if the quarto extension could warn about things like invalid quarto specific syntax like the wrong include shortcode.

**Reply by mcanouil** [reply]
> Why not? If something is not in path it does not exist?!
> E.g. in python linters warn when undefined variables are detected.

Your are mixing things up here. Paths are not variables and here there are no variables.
A shortcode is plain text which in the case of Quarto implies `{{< >}}` syntax but does not require this as it's parsed by Lua.
Linters do static code analysis, they are not aware of the file system and does not evaluate code.
Linters check syntax not code validity/evaluability.
Quarto offers the ability to anyone to create Lua filter that could make `{{< >>>}}` valid. Again, linters don't evaluate anything.

> It would be nice if the quarto extension could warn about things like invalid quarto specific syntax like the wrong include shortcode.

See above why it's not that simple.
